// gei-backend/dist/config/database.config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.dataSourceConfig = exports.databaseConfig = void 0;
const databaseConfig = (configService) => ({
    type: 'postgres',
    host: configService.get('DATABASE_HOST') || 'localhost',
    port: configService.get('DATABASE_PORT') || 5432,
    username: configService.get('DATABASE_USERNAME') || 'postgres',
    password: configService.get('DATABASE_PASSWORD') || 'password',
    database: configService.get('DATABASE_NAME') || 'gei_db',
    entities: [__dirname + '/../**/*.entity{.ts,.js}'],
    migrations: [__dirname + '/../database/migrations/*{.ts,.js}'],
    synchronize: configService.get('NODE_ENV') === 'development',
    logging: configService.get('NODE_ENV') === 'development',
    ssl: configService.get('DATABASE_SSL') ? { rejectUnauthorized: false } : false,
    autoLoadEntities: true,
    keepConnectionAlive: true,
});
exports.databaseConfig = databaseConfig;
const dataSourceConfig = (configService) => ({
    type: 'postgres',
    host: configService.get('DATABASE_HOST') || 'localhost',
    port: configService.get('DATABASE_PORT') || 5432,
    username: configService.get('DATABASE_USERNAME') || 'postgres',
    password: configService.get('DATABASE_PASSWORD') || 'password',
    database: configService.get('DATABASE_NAME') || 'gei_db',
    entities: [__dirname + '/../**/*.entity{.ts,.js}'],
    migrations: [__dirname + '/../database/migrations/*{.ts,.js}'],
    synchronize: configService.get('NODE_ENV') === 'development',
    logging: configService.get('NODE_ENV') === 'development',
    ssl: configService.get('DATABASE_SSL') ? { rejectUnauthorized: false } : false,
});
exports.dataSourceConfig = dataSourceConfig;
